### PR TITLE
Add helper functions to make tool usable n times

### PR DESCRIPTION
**Pull request by Wuzzy2**
This adds two functions:

* `item:add_wear_by_uses`
* `minetest.get_tool_wear_after_use`

These are two helper functions for tools that have a specified number of uses before breaking. Two examples for such tools are MTG's Screwdriver and Flint and Steel. This makes sure the tool will break after the exact specified number of uses.

The first function actually adds the wear to the tool and (if neccessary) breaks it.
The second function only "simulates" it, analog to `get_dig_params` and `get_hit_params`.

It also fixes a minor mistake in the function comments. `add_wear` claims to support only a range of [0,65535] but this is false, it also supports 65536 (which leads to insta-break). I checked this.

## Context
The common (and naive) way that mods usually implement "tool break after *n* uses` is by adding wear with `tool:add_wear(some_fixed_value)` in which `some_fixed_value` is chosen in a way that the tool breaks after exactly `n` times. This works for small numbers of `n` (<273) but as `n` increases you might actually not be able to hit `n` exactly due to rounding. As an extreme example, with this method you can't have a tool with exactly 50000 uses, but must decide 32768 or 65536 uses (adding 2 or 1 wear at every use, respectively).
That's where the helper function comes into play.

This PR builds on the work done in #11110 and is basically an extension of the logic used there for mods. It uses `calculateResultWear` for the calcuation. Read this PR for more context.

Note the actual math is NOT altered by this PR in any way. It basically just exposes the "tool uses before it breaks" calculation the engine already uses to Lua.

## To do

This PR is ready for review.

## How to test

* Activate `luacmd`
* Give yourself any tool with 0 wear and wield it
* `/lua n=60000; p=0; x=me:get_wielded_item(); for i=1,65536 do x:add_wear_by_uses(n); p=p+1; if x:get_count()==0 then break end end print(p); me:set_wielded_item(x)`
* This is a loop that calls `add_wear_by_uses(n)` `n` times, so the tool must break after exactly `n` uses. At the end, it prints out how many iterations this look actually had
* Since `n` is 60000 in this case, it should print out 60000
* Replace the value of `n` with different values, including 0 (=infinite uses), 65535 or 65536
* The other function can be tested in a similar fashion
* Also try manually calling these functions without a loop to see how the wear bar reacts

### CODE REVIEW

**Comment by runsy** [comment]
This would be cool.

**Comment by Df458** [comment]
Tested with all integers between [0 .. 65537]
- 0 and 65537 never broke
- All other values broke in the expected number of uses

At a glance, behavior seems correct. Will test more and finish review tomorrow.

**Review by Df458 — APPROVED** [review]
Looked over the code, also did some further manual testing mixing `uses` values and behavior seemed sensible. Looks good to me.

File: src/script/lua_api/l_item.cpp [hunk 1]
```diff
@@ -356,6 +356,26 @@ int LuaItemStack::l_add_wear(lua_State *L)
...
+int LuaItemStack::l_add_wear_by_uses(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	LuaItemStack *o = checkobject(L, 1);
+	ItemStack &item = o->m_stack;
+	u32 uses = readParam<int>(L, 2);
+	u16 initial_wear = item.wear;
+	u32 add_wear = calculateResultWear(uses, initial_wear);
+	bool result = item.addWear(add_wear, getGameDef(L)->idef());
+	lua_pushboolean(L, result);
```
Comment (sfan5) [inline src/script/lua_api/l_item.cpp:375]:
```suggestion
	u32 add_wear = calculateResultWear(uses, item.wear);
	bool result = item.addWear(add_wear, getGameDef(L)->idef());
	lua_pushboolean(L, result);
```

File: src/script/lua_api/l_util.cpp [hunk 2]
```diff
@@ -159,6 +159,19 @@ int ModApiUtil::l_write_json(lua_State *L)
...
 }
 
+// get_tool_wear_after_use(uses[, initial_wear])
+int ModApiUtil::l_get_tool_wear_after_use(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	u32 uses = readParam<int>(L, 1);
+	u16 initial_wear = 0;
+	if (!lua_isnoneornil(L, 2))
+		initial_wear = readParam<int>(L, 2);
```
Comment (sfan5) [inline src/script/lua_api/l_util.cpp:169]: I believe this works too
```suggestion
	u16 initial_wear = readParam<int>(L, 2, 0);
```

File: src/script/lua_api/l_item.h [hunk 3]
```diff
@@ -108,11 +108,20 @@ class LuaItemStack : public ModApiBase {
...
 	// Returns true if the item is (or was) a tool.
 	static int l_add_wear(lua_State *L);
 
+	// add_wear_by_uses(self, uses) -> true/false
+	// The range for "uses" is [0,65536].
+	// Adds wear to the item in such a way that, if
+	// only this function is called to add wear, the item
+	// will be destroyed exactly after `uses` times of calling it.
+	// No-op if `uses` is 0 or item is not a tool.
+	// Returns true if the item is (or was) a tool.
```
Comment (sfan5) [inline src/script/lua_api/l_item.h:122]: just as information: I don't think it's necessary to put detailed documentation here, btw

**Comment by sfan5** [comment]
If you have time it'd also be nice to see [unittests in devtest](https://github.com/minetest/minetest/tree/master/games/devtest/mods/unittests) for this API function.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -6375,7 +6381,13 @@ an itemstring, a table or `nil`.
   or those of the hand if none are defined for this item type
 * `add_wear(amount)`
     * Increases wear by `amount` if the item is a tool, otherwise does nothing
+    * Valid `amount` range is [0,65536]
     * `amount`: number, integer
+* `add_wear_by_uses(uses)`
```
Comment (rubenwardy) [inline doc/lua_api.txt:6386]: I feel like it would make more sense if it's called `max_uses` rather than `uses`

File: doc/lua_api.txt [hunk 5]
```diff
@@ -3512,6 +3512,12 @@ Helper functions
 * `minetest.pointed_thing_to_face_pos(placer, pointed_thing)`: returns a
   position.
     * returns the exact position on the surface of a pointed node
+* `minetest.get_tool_wear_after_use(uses [, initial_wear])`
+    * Simulates a tool being used once and returns the added wear,
+    * such that, if only this function is used to calculate wear,
```
Comment (rubenwardy) [inline doc/lua_api.txt:3517]: Broken bullet points. Should only have a `*` on new bullet points, other lines should omit the * but line up the text

**Comment by Wuzzy2** [comment]
I fixed the broken bullet points and I renamed the arg.

**Review by sfan5 — APPROVED** [review]
Looks good.

**Pull request merged**